Enhance OIDC Discovery Provider to better support multiple domains

marcofranssen opened the pull request:
This PR builds on top of #2398 

It adds the capability to host this component on both internal and external endpoints. See [Helm chart](https://github.com/philips-labs/helm-charts/pull/24)

Also updated the README for an undocumented config option. 

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**

URLs in `/.well-known/openid-configuration`.

**Description of change**

The URLs match the scheme the oidc discovery provider is hosted on. Also when running behind a reverse proxy like Nginx HAProxy Traefik. In addition it dynamically takes the domain to be put in the response of the `/.well-known/openid-configuration`.

This allows us to expose the service on an ingress as seen in the helm chart, but also to have other services in the kube-cluster use the internal DNS entry.

e.g.

http://spire-oidc.spire (internal)
https://oidc-discovery.mydomain.com (external)

**Which issue this PR fixes**

N.A.

**Related to**:

https://github.com/philips-labs/helm-charts/pull/24
#2398

marcofranssen commented:
Rebased the PR on latest main.

azdagron commented:
Thanks for opening this, @marcofranssen! Since this subsumes #2398, would it make sense to close #2398 and hold a discussion here around how we populate the host and scheme for the URLs produced by the provider?

From what I can tell you want to enable the following:
1.  ability to host the provider behind a proxy that handles TLS termination
2.  ability to host the provider behind multiple DNS names (e.g. internal and external)

Both of those seem like reasonable things for the provider to support.

One thing that worries me about the proposed changes is that it is becomes possible to serve the resources over an HTTP URL. Due to the security sensitive nature of these resources, they should never be served over HTTP, even internally, and is a practice we've actively discouraged by hardcoding `https` even when the provider itself doesn't terminate TLS.  I don't think that relying on the `X-Forwarded-Scheme` header is the right choice either. A broken proxy that fails to send the header would downgrade the security posture of the bundle endpoint. My preference would be to maintain a hard-coded `https` scheme for the URLs produced by the provider. If that isn't going to work out, my second preference would be to wire up a new configurable, e.g. `insecure_scheme = true` that would produce `http` URLs, which should only be enabled in local developer setups, if ever. If that option is enabled, we should log very aggressively that the setup is insecure to maybe mitigate it being accidentally turned on in production.

I have some additional comments on the code, but I'll wait to post those until we come to agreement on the approach, since some will no longer be applicable depending on the route we go.

marcofranssen commented:
@azdagron correct. Intention was to be able to deploy at internal and external endpoints from the Helm chart. I have updated the PR to always redirect http to https endpoints, unless the `allow_insecure_scheme` is defined. Therefore I made `insecure_addr` an optional config which defaults to `:80`.

marcofranssen commented:
With the latest changes I am successfully able to deploy the Helm chart on both my local and eks environments.

https://github.com/philips-labs/helm-charts/pull/24

From the Helm chart I dropped the ingress and changed to use a Loadbalancer Service or NodePort to expose the service out of the cluster. This is necessary to make the HostSNI challenge work with acme.

Currently on local I'm disabling the `http->https` redirect by setting the flag. On EKS I enable the `http->https` redirect (default in the chart). Added value of this `http->https` redirect is some improved user experience.

See here the values files I'm using for EKS and local deploy with this chart and a custom build of the spire-oidc image (build with changes from this branch).

<details>
  <summary>helm-eks-values.yml</summary>

```yaml
imagePullSecrets:
  - name: my-registry
    username: "$MY_REGISTRY_USER"
    password: "$MY_REGISTRY_PASSWORD"
    registryURL: my.registry.com

server:
  dataStorage:
    enabled: false

oidc:
  enabled: true

  image:
    repository: my.registry.com/dev/spire-oidc-discovery-provider
    tag: latest
    pullPolicy: Always

  logLevel: DEBUG

  allowInsecureScheme: false

  service:
    type: LoadBalancer
    annotations:
      external-dns.alpha.kubernetes.io/hostname: dev-oidc-spire.my-eks-environment.com

  domain: dev-oidc-spire.my-eks-environment.com

  acme:
    tosAccepted: true
    emailAddress: marco.franssen@my-email.com

spire:
  clusterName: "awesomeness-cluster"
  trustDomain: "dev.my-eks-environment.com"
  agent:
    logLevel: INFO
  server:
    logLevel: INFO
```
</details>


<details>
  <summary>helm-local-values.yml</summary>

```yaml
server:
  dataStorage:
    enabled: false

oidc:
  enabled: true

  image:
    repository: oidc-discovery-provider
    tag: latest

  logLevel: DEBUG

  allowInsecureScheme: true

  acme:
    tosAccepted: true
    emailAddress: marco.franssen@my-email.com
    directoryUrl: https://acme-staging-v02.api.letsencrypt.org/directory

  domain: oidc-spire.dev.localhost

spire:
  clusterName: "dev-cluster"
  trustDomain: "dev.localhost"
  agent:
    logLevel: DEBUG
  server:
    logLevel: DEBUG
```
<details>

Please let me know if most recent changes are sufficient to get it merged. I might plan as well in the future to deploy `step-ca` on my local cluster that would allow me to issue acme certs on private domains.

We might want to file an issue though to also work on an improvement to run on https via spire issued certs or other means of running on TLS. That involves more changes though which are not required at this stage to make the Helm chart working.

marcofranssen commented on support/oidc-discovery-provider/handler.go:
```diff
@@ -34,14 +44,31 @@ func (h *Handler) serveWellKnown(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	if h.allowInsecureScheme {
```
this only serves the urls as HTTP if allowInsecureScheme is configured and if the origin request is not served on HTTPS. So in allowInsecure is enabled but the Reverse Proxy runs on HTTPS, it is still served on HTTPS.

marcofranssen commented on support/oidc-discovery-provider/handler.go:
```diff
@@ -34,14 +44,31 @@ func (h *Handler) serveWellKnown(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if !h.domains[r.Host] {
```
Only serves content when requested via a preconfigured domain, this eleminates rougue proxies or other tricks in the infrastructure.

marcofranssen commented on support/oidc-discovery-provider/README.md:
```diff
@@ -30,18 +30,20 @@ The provider has the following command line flags:
+| ----------------------  | --------| ----------- | -------------------------------------------------------- | -------- |
+| `acme`                  | section | required[1] | Provides the ACME configuration.                         |          |
+| `domain`                | string  | required    | The domain the provider is being served from.            |          |
+| `insecure_addr`         | string  | optional    | Exposes the service on http.                             | `":80"`  |
```
Let me know if I should leave these 2 settings undocumented.

azdagron replied:
I am fine either way. If we do leave them documented, we should at a minimum add an annotation down below (e.g. `[2]` that explains that at least the `allow_insecure_scheme` configurable is for test purposes only and should _never_ be used in a production deployment.

@evan2645 may have stronger feelings around the documentation of these.

marcofranssen replied:
added https://github.com/spiffe/spire/pull/2404/files#diff-30140922f86d99ba52fbcc88de127aa485375d6c66564a6f91af994c733e315bR52

@azdagron @evan2645

azdagron commented on support/oidc-discovery-provider/handler.go:
```diff
@@ -34,14 +44,31 @@ func (h *Handler) serveWellKnown(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	if h.allowInsecureScheme {
+		if r.URL.Scheme == "" {
+			if r.TLS == nil {
+				r.URL.Scheme = "http"
+			} else {
+				r.URL.Scheme = "https"
+			}
+		}
+	} else {
+		r.URL.Scheme = "https"
+	}
```
nit: can be simplified for readability:
```suggestion
	urlScheme := "https"
	if h.allowInsecureScheme && r.URL.Scheme == "" && r.TLS == nil {
		urlScheme = "http"
	}
```
Followed by the use of `urlScheme` when constructing the URLs below.

azdagron commented on support/oidc-discovery-provider/handler.go:
```diff
@@ -34,14 +44,31 @@ func (h *Handler) serveWellKnown(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if _, ok := h.domains[r.Host]; !ok {
+		http.Error(w, "domain not allowed", http.StatusNotFound)
+		return
+	}
```
I don't think we can do this unconditionally without breaking existing deployments who have not configured their reverse proxies to add the `X-Forwarded-Host` header.

What if we only enable this validation if the `Domains` configurable was used and maintain existing behavior if `Domain` was used instead? When we finally deprecate `Domain`, we'll be in a stronger position to be doing this validation for every deployment, since those who have been using (or have migrated to) `Domains` will already have had to adjust their proxy deployment to add the header.

marcofranssen replied:
Done.

azdagron commented on support/oidc-discovery-provider/config.go:
```diff
@@ -23,10 +23,17 @@ type Config struct {
 	// LogRequests is a debug option that logs all incoming requests
 	LogRequests bool `hcl:"log_requests"`
 
-	// Domain is the domain this provider will be hosted under. It is used
-	// as the domain when building the JWKS URI. It is also used when obtaining
-	// obtaining certs via ACME (unless InsecureAddr is specified).
-	Domain string
```
We need to keep the old configurable or we will break existing deployments on upgrade. I think the right path forward is add a warning log if the `domain` configurable is used saying something like:

```
The domain configurable is deprecated and will be removed in a future release. Please use the domains configurable instead.
```

We should fail if both `domains` and `domain` are set.

If `domain` is set, I think we should avoid the new domain validation that was added to the handler. This gives us a pretty good migration story towards strong domain validation without breaking existing deployments.

marcofranssen replied:
Done…

If domain is set we skip the domain validation in the handler.

If domain setting is configured we will copy the value in the Domains setting so all logic can be based on the new setting, to reduce amount of refactors when moving the deprecated setting out.

azdagron commented on support/oidc-discovery-provider/config.go:
```diff
@@ -224,6 +241,20 @@ func ParseConfig(hclConfig string) (_ *Config, err error) {
 
+func unique(items []string) []string {
+	keys := make(map[string]bool)
+	list := make([]string, 0)
```
Allocating a zero length slice is a little confusing unless there is a requirement that the function never return a `nil` slice, which isn't typical. I'd rather see:
```
var list []string
```
or
```
list := make([]string, 0, len(items))
```

azdagron commented on support/oidc-discovery-provider/config.go:
```diff
@@ -143,8 +155,13 @@ func ParseConfig(hclConfig string) (_ *Config, err error) {
+
+	if c.Domain != "" {
+		c.Domains = append([]string{c.Domain}, c.Domains...)
+		c.Domains = unique(c.Domains)
```
Not that it really matters, since they are just put into a set at a later point, but should we run `unique` either way? Nothing keeps folks from declaring the same domain more than once in the `Domains` configurable.

marcofranssen replied:
It makes it at least more predictable for future code that uses the domains. Also covered this in a testcase.

azdagron replied:
In that case, we should dedup `c.Domains` whether `c.Domain` was appended to it (i.e. move the dedup call outside, and after, the if block).

azdagron commented on support/oidc-discovery-provider/main.go:
```diff
@@ -39,13 +39,17 @@ func run(configPath string) error {
 	defer log.Close()
 
+	if config.Domain != "" {
+		log.Warn("`domain` configuration is deprecated, use domains configuration instead.")
```
SPIRE uses documentation refers to these as "configurables" :)
```suggestion
		log.Warn("The `domain` configurable is deprecated and will be removed in a future release; use `domains` instead.")
```

azdagron commented on support/oidc-discovery-provider/handler.go:
```diff
@@ -34,14 +46,30 @@ func (h *Handler) serveWellKnown(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Scheme == "" && r.TLS == nil {
+			urlScheme = "http"
+		} else {
+			urlScheme = r.URL.Scheme
```
AFAICT the scheme won't be set on non-proxied requests (or requests that omit the header). So if somebody sets allow_insecure_scheme and self-terminates with ACME, the urlScheme will end up empty. I think we should only be overriding `https` if the URL scheme is not set (implying we connected directly and not through a proxy) and we aren't terminating TLS, right?

marcofranssen replied:
That breaks my changes to run on the socket_path and proxy via Nginx over http. In my case I am setting the url scheme accordingly in the proxy.

marcofranssen replied:
Simplified the check to work in all scenarios.

azdagron reviewed (commented):
We're getting close :) Mostly nitpicks but a few comments of substance...

azdagron reviewed (approved):
Thanks, @marcofranssen!
